pkg/dhcp6client/client.go

```diff
@@ -0,0 +1,499 @@
+	timeout time.Duration
+}
+
+// New returns a new DHCPv6 client based on the given parameters.
+func New(iface netlink.Link, opts ...ClientOpt) (*Client, error) {
+	haddr := iface.Attrs().HardwareAddr
+	ip, err := eui64.ParseMAC(net.ParseIP("fe80::"), haddr)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := net.ListenUDP("udp6", &net.UDPAddr{
+		IP:   ip,
+		Port: ClientPort,
+		Zone: iface.Attrs().Name,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Client{
+		iface:   iface,
+		conn:    conn,
+		timeout: 10 * time.Second,
+		retry:   3,
+	}
+
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+	return c, nil
+}
+
+// ClientOpt is a function that configures the client.
+type ClientOpt func(*Client) error
+
+// WithTimeout configures the retransmission timeout.
+//
+// Default is 10 seconds.
+//
+// TODO(hugelgupf): Check RFC for retransmission behavior.
+func WithTimeout(d time.Duration) ClientOpt {
+	return func(c *Client) error {
+		c.timeout = d
+		return nil
+	}
+}
+
+// WithRetry configures the retransmission counts.
+//
+// Default is 3.
+//
+// TODO(hugelgupf): Check RFC for retransmission behavior.
+func WithRetry(retry int) ClientOpt {
+	return func(c *Client) error {
+		c.retry = retry
+		return nil
+	}
+}
+
+// RapidSolicit solicits one non-temporary address assignment by multicasting a
+// DHCPv6 solicitation message with the rapid commit option.
+//
+// RapidSolicit returns the first valid, suitable response by any remote server.
+func (c *Client) RapidSolicit() (*dhcp6opts.IANA, *dhcp6.Packet, error) {
+	solicit, err := NewRapidSolicit(c.iface.Attrs().HardwareAddr)
+	if err != nil {
+		return nil, nil, err
+	}
+	return c.RequestOne(solicit)
+}
+
+// RequestOne multicasts the `request` and returns the first matching IANA and
+// its associated Packet returned by any server.
+func (c *Client) RequestOne(request *dhcp6.Packet) (*dhcp6opts.IANA, *dhcp6.Packet, error) {
+	ianas, pkt, err := c.Request(request)
+	if err != nil {
+		return nil, nil, err
+	}
+	if len(ianas) != 1 {
+		return nil, nil, fmt.Errorf("got %d IANAs, expected 1", len(ianas))
+	}
+	return ianas[0], pkt, nil
+}
+
+// Solicit multicasts a Solicit message and collects all Advertise responses
+// received before c.timeout expires.
+//
+// Solicit blocks until either:
+// - `ctx` is canceled; or
+// - we have exhausted all configured retries and timeouts.
+func (c *Client) Solicit(ctx context.Context) ([]*dhcp6.Packet, error) {
+	solicit, err := NewSolicitPacket(c.iface.Attrs().HardwareAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	out, errCh := c.SimpleSendAndRead(ctx, DefaultServers, solicit)
+
+	var ads []*dhcp6.Packet
+	// resps is closed by SimpleSendAndRead when done.
+	for r := range out {
+		if r.Packet.MessageType == dhcp6.MessageTypeAdvertise {
+			ads = append(ads, r.Packet)
+		}
+	}
+
+	if err, ok := <-errCh; ok && err != nil {
+		return nil, err
+	}
+	return ads, nil
+}
+
+// This name smells.
```

Comment:
LOL :D

something like `multiErrs`, `errorGroup`, `errorList` or similar? None is really good tbh

Reply:
done